nucliadb/nucliadb/export_import/consumers.py

```diff
@@ -0,0 +1,270 @@
+            queue=group,
+            stream=stream,
+            flow_control=True,
+            cb=self.subscription_worker,
+            subscription_lost_cb=self.setup_nats_subscription,
+            config=nats.js.api.ConsumerConfig(
+                deliver_policy=nats.js.api.DeliverPolicy.BY_START_SEQUENCE,
+                opt_start_seq=1,
+                ack_policy=nats.js.api.AckPolicy.EXPLICIT,
+                max_ack_pending=nats_consumer_settings.nats_max_ack_pending,
+                max_deliver=nats_consumer_settings.nats_max_deliver,
+                ack_wait=nats_consumer_settings.nats_ack_wait,
+                idle_heartbeat=nats_consumer_settings.nats_idle_heartbeat,
+            ),
+        )
+        logger.info(f"Subscribed to {subject} on stream {stream}")
+
+    async def subscription_worker(self, msg: Msg):
+        subject = msg.subject
+        reply = msg.reply
+        seqid = int(reply.split(".")[5])
+        logger.info(
+            f"Message received: subject:{subject}, seqid: {seqid}, reply: {reply}"
+        )
+        async with MessageProgressUpdater(
+            msg, nats_consumer_settings.nats_ack_wait * 0.66
+        ):
+            try:
+                await self.do_work(msg)
+            except UNRECOVERABLE_ERRORS as e:
+                # Unrecoverable errors are not retried
+                errors.capture_exception(e)
+                await msg.ack()
+            except Exception as e:
+                # Unhandled errors are retried after a small delay
+                errors.capture_exception(e)
+                await asyncio.sleep(2)
+                await msg.nak()
+                raise e
+            else:
+                # Successful messages are acked
+                await msg.ack()
+
+    async def do_work(self, msg: Msg):
+        raise NotImplementedError()
+
+
+class ExportsConsumer(BaseConsumer):
+    stream = const.Streams.EXPORTS
+
+    async def do_work(self, msg: Msg):
+        export_msg = ExportMessage.parse_raw(msg.data)
+        export_id = export_msg.export_id
+        kbid = export_msg.kbid
+        metadata = await self.data_manager.get_export_metadata(kbid, export_id)
+
+        if metadata.status in (Status.FINISHED, Status.ERRORED):
+            logger.info(f"Export {export_id} for {kbid} is {metadata.status.value}.")
+            return
+
+        if metadata.tries >= EXPORT_MAX_RETRIES:
+            logger.info(
+                f"Export {export_id} for kbid {kbid} failed too many times. Skipping"
+            )
+            metadata.status = Status.ERRORED
+            await self.set_metadata(metadata)
+            return
+
+        if metadata.status == Status.FAILED:
+            logger.info(f"Export {export_id} for kbid {kbid} failed, retrying")
+            metadata.status = Status.RUNNING
+        else:  # metadata.status == Status.SCHEDULED:
+            logger.info(f"Starting export {export_id} for kbid {kbid}")
+            metadata.status = Status.RUNNING
```

Comment:
about all this task-related logic, I thought about creating a `async_tasks` top level component in nucliadb that would offer an api to register a function to a particular stream. thoughts?

Reply:
something like:

```python
export_consumer = TaskConsumer.consume_stream(
    stream=const.Stream.EXPORT,
    func=export_kb,
    message_type=ExportMessage,
    partitions=[1],
) 

```

Reply:
I like it. But remove the partitioning as that is something unusual and not a normal pattern.

I would then like to move some other things to it that I did with pub/sub that I screwed up and don't think should actually be pubsub anymore(because of how concurrency works)